File: backend/src/main/java/com/unina/ratatouille23/menu/entity/Categoria.java
package com.unina.ratatouille23.menu.entity;

import java.util.List;

import com.fasterxml.jackson.annotation.JsonManagedReference;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.OneToMany;

@Entity
public class Categoria {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    @Column(unique = true)
    private int id;

    @Column(nullable = false)
    private String nome;

    @Column(nullable = false)
    private int idRistorante;

    @OneToMany(mappedBy = "categoria", cascade = CascadeType.ALL, orphanRemoval = true)
    @JsonManagedReference
    private List<Elemento> elementiDellaCategoria;

    
    public Categoria() {
    }

    public int getId() {
        return this.id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public String getNome() {
        return this.nome;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public int getIdRistorante() {
        return this.idRistorante;
    }

    public void setIdRistorante(int idRistorante) {
        this.idRistorante = idRistorante;
    }

    public List<Elemento> getElementiDellaCategoria() {
        return this.elementiDellaCategoria;
    }

    public void setElementiDellaCategoria(List<Elemento> elementiDellaCategoria) {
        this.elementiDellaCategoria = elementiDellaCategoria;
    }


}

File: client/lib/models/menu/Categoria.dart
import 'Elemento.dart';

class Categoria{

  late int _id;
  late String _nome;
  List<Elemento>? _elementi;
  late int _id_ristorante;

  Categoria.senzaIdAndElementi(this._nome, this._id_ristorante);

  Categoria(int id, String nome, List<Elemento> elementi, int id_ristorante) {
    _id = id;
    _nome = nome;
    _elementi = elementi;
    _id_ristorante=id_ristorante;
  }

  Categoria.senzaElementi(int id, String nome, int id_ristorante) {
    _id = id;
    _nome = nome;
    _id_ristorante=id_ristorante;
  }

  Categoria.fromJson(Map<String, dynamic> json) {
    _id=json['id'];
    _nome=json['nome'];
    _id_ristorante=json['idRistorante'];
    _elementi = [];
    json['elementiDellaCategoria'].forEach((v) {
      _elementi?.add(Elemento.fromJson(v));
    });

  }

  Map<String, dynamic> toJsonSenzaId(){
    return{
      'nome':_nome,
      'idRistorante':_id_ristorante
    };
  }

  Map<String, dynamic> toJson(){
    return{
      'id':_id,
      'nome':_nome,
      'idRistorante':_id_ristorante
    };
  }
    void set_id(int id){_id=id;}
    void set_nome(String nome){_nome=nome;}
    void set_elementi(List<Elemento> elemento){_elementi=elemento;}

    int get_id(){return _id;}
    String get_nome(){return _nome;}
    List<Elemento>? get_elementi(){return _elementi;}
    int getIDRistorante(){return _id_ristorante;}

}
